### PR TITLE
Infer into recursive mapped type targets

### PR DESCRIPTION
**Pull request by Andarist**
fixes https://github.com/microsoft/TypeScript/issues/53650

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -2064,7 +2064,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
     /** Key is "/path/to/a.ts|/path/to/b.ts". */
     var amalgamatedDuplicates: Map<string, DuplicateInfoForFiles> | undefined;
     var reverseMappedCache = new Map<string, Type | undefined>();
-    var inInferTypeForHomomorphicMappedType = false;
```
Comment (Andarist) [inline src/compiler/checker.ts:2067]: the flag was "global" and thus prevented recursing here just for any type even if `source` was different or if we'd encounter different `target`s

**Review by weswigham — APPROVED** [review]
Assuming perf/extended suites are fine, I like this, iirc there was just a test/project which crashed without some kind of stop like this originally, and the globally flag was simply expedient - definitely not complete.

**Comment by weswigham** [comment]
@typescript-bot test this
@typescript-bot run dt
@typescript-bot test top100
@typescript-bot perf test this

**Comment by jakebailey** [comment]
@typescript-bot test this
@typescript-bot test top100
@typescript-bot user test this
@typescript-bot user test tsserver
@typescript-bot test tsserver top100
@typescript-bot run dt
@typescript-bot perf test this
@typescript-bot pack this

**Comment by jakebailey** [comment]
Is there anything left here? This is approved and seems to be fine to my eye.

**Comment by Andarist** [comment]
> Is there anything left here?

Yes, merging it 😜

**Comment by jakebailey** [comment]
Just looking back at this, the variable mentioned above is from #38224; presumably, the test case from that PR still exists and is passing and I haven't been able to find a linked project which shows a perf issue being fixed by it, so I guess things seem safe?

I'm not the assignee so I didn't want to click the green button, but I'm not sure how pressing / safe this one is.

**Comment by Andarist** [comment]
I definitely won't argue that this is a pressing PR so if you don't feel like merging this, don't feel pressured to do it. I'd consider this one to be on the safer side of things though (especially with Wes' approval) 😉

**Comment by ssalbdivad** [comment]
@jakebailey My total lack of context on the underlying changes notwithstanding, I'd still like to take the opportunity to express my appreciation of this work and hopes that the PR will be merged so I can also close the [associated issue](https://github.com/arktypeio/arktype/issues/565) in ArkType !🙏

**Comment by jakebailey** [comment]
I think everyone on the review list is out; I'm gonna click the green button so we can get this tested pre-RC (and it's also a bugfix that does not seem to show any regression in extended testing); if it is a problem I'll "unclick" the green button.

**Pull request merged**